src/connector/src/sink/coordinate.rs

```diff
@@ -0,0 +1,153 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::time::{Instant, SystemTime, UNIX_EPOCH};
+
+use anyhow::anyhow;
+use chrono::Utc;
+use risingwave_common::array::StreamChunk;
+use risingwave_common::buffer::Bitmap;
+use risingwave_pb::connector_service::sink_writer_to_coordinator_msg::{
+    CommitRequest, StartCoordinationRequest,
+};
+use risingwave_pb::connector_service::{
+    sink_coordinator_to_writer_msg, sink_writer_to_coordinator_msg, SinkCoordinatorToWriterMsg,
+    SinkMetadata, SinkWriterToCoordinatorMsg,
+};
+use risingwave_rpc_client::{BidiStreamHandle, SinkCoordinationRpcClient};
+use tracing::{debug, info, warn};
+
+use crate::sink::{Result, SinkError, SinkParam, SinkWriter};
+
+struct CoordinatorStreamHandle(
+    BidiStreamHandle<SinkWriterToCoordinatorMsg, SinkCoordinatorToWriterMsg>,
+);
+
+impl CoordinatorStreamHandle {
+    async fn new(
+        mut client: SinkCoordinationRpcClient,
+        param: SinkParam,
+        vnode_bitmap: Bitmap,
+    ) -> Result<Self> {
+        let (stream_handle, first_response) = BidiStreamHandle::initialize(
+            SinkWriterToCoordinatorMsg {
+                msg: Some(sink_writer_to_coordinator_msg::Msg::StartRequest(
+                    StartCoordinationRequest {
+                        vnode_bitmap: Some(vnode_bitmap.to_protobuf()),
+                        param: Some(param.to_proto()),
+                    },
+                )),
+            },
+            |req_stream| async move { client.coordinate(req_stream).await },
+        )
+        .await?;
+        match first_response {
+            SinkCoordinatorToWriterMsg {
+                msg: Some(sink_coordinator_to_writer_msg::Msg::StartResponse(_)),
+            } => Ok(Self(stream_handle)),
+            msg => Err(SinkError::Coordinator(anyhow!(
+                "should get start response but get {:?}",
+                msg
+            ))),
+        }
+    }
+
+    async fn commit(&mut self, epoch: u64, metadata: SinkMetadata) -> Result<()> {
+        self.0
+            .send_request(SinkWriterToCoordinatorMsg {
+                msg: Some(sink_writer_to_coordinator_msg::Msg::CommitRequest(
+                    CommitRequest {
+                        epoch,
+                        metadata: Some(metadata),
+                    },
+                )),
+            })
+            .await?;
+        match self.0.next_response().await? {
+            SinkCoordinatorToWriterMsg {
+                msg: Some(sink_coordinator_to_writer_msg::Msg::CommitResponse(_)),
+            } => Ok(()),
+            msg => Err(SinkError::Coordinator(anyhow!(
+                "should get commit response but get {:?}",
+                msg
+            ))),
+        }
+    }
+}
+
+pub struct CoordinatedSinkWriter<W: SinkWriter<CommitMetadata = Option<SinkMetadata>>> {
```

Comment:
Seems no associated type in the `SinkWriter`, why need `CommitMetadata` here?

Reply:
Details are in https://github.com/risingwavelabs/risingwave/pull/10757.

In short, the definition of `SinkWriter` and `CoordinatedSinkWriter` are very similar and only differ in that the return type `barrier` of `CoordinatedSinkWriter` is `Option<SinkMetadata>`. Therefore, in https://github.com/risingwavelabs/risingwave/pull/10757, I introduce a associated type `CommitMetadata` so that the two trais can share most of the method definition.